src/source-dpdk.c

```diff
@@ -0,0 +1,525 @@
+            last_dump = current_time;
+        }
+    }
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+/**
+ * \brief Init function for ReceiveDPDK.
+ *
+ * \param tv pointer to ThreadVars
+ * \param initdata pointer to the interface passed from the user
+ * \param data pointer gets populated with DPDKThreadVars
+ *
+ */
+TmEcode ReceiveDPDKThreadInit(ThreadVars *tv, const void *initdata, void **data)
+{
+    SCEnter();
+    int retval;
+    DPDKIfaceConfig *dpdk_config = (DPDKIfaceConfig *)initdata;
+
+    if (initdata == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "DPDK Configuration failed");
+        dpdk_config->DerefFunc(dpdk_config);
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    DPDKThreadVars *ptv = SCCalloc(1, sizeof(DPDKThreadVars));
+    if (unlikely(ptv == NULL)) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Unable to allocate memory");
+        dpdk_config->DerefFunc(dpdk_config);
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    ptv->tv = tv;
+    ptv->pkts = 0;
+    ptv->bytes = 0;
+    ptv->livedev = LiveGetDevice(dpdk_config->iface);
+    DPDKSetTimevalOfMachineStart(&ptv->machine_start_time);
+
+#ifdef PACKET_STATISTICS
+    ptv->capture_dpdk_packets = StatsRegisterCounter("capture.packets", ptv->tv);
+    ptv->capture_dpdk_drops = StatsRegisterCounter("capture.drops", ptv->tv);
+#endif
+
+    ptv->copy_mode = dpdk_config->copy_mode;
+    ptv->checksum_mode = dpdk_config->checksum_mode;
+
+    ptv->threads = dpdk_config->threads;
+    ptv->port_id = dpdk_config->port_id;
+    ptv->out_port_id = dpdk_config->out_port_id;
+    uint16_t queue_id = SC_ATOMIC_ADD(dpdk_config->queue_id, 1);
+    ptv->queue_id = queue_id;
+    // pass the pointer to the mempool and then forget about it. Mempool is freed in thread deinit.
+    ptv->pkt_mempool = dpdk_config->pkt_mempool_array[queue_id];
+    dpdk_config->pkt_mempool_array[queue_id] = NULL;
+
+    // the last thread starts the device
+    if (queue_id == dpdk_config->threads - 1) {
+        retval = rte_eth_dev_start(ptv->port_id);
+        if (retval < 0) {
+            SCLogError(SC_ERR_DPDK_INIT, "Error (err=%d) during device startup of port %u", retval,
+                    ptv->port_id);
+            dpdk_config->DerefFunc(dpdk_config);
+            SCReturnInt(TM_ECODE_FAILED);
+        }
+    }
+
+    if (rte_eth_dev_socket_id(ptv->port_id) != (int)rte_socket_id()) {
+        SCLogWarning(SC_WARN_DPDK_CONF,
+                "NIC on a different NUMA node as lcore %u is initialized on", rte_socket_id());
+    }
+
+    *data = (void *)ptv;
+    dpdk_config->DerefFunc(dpdk_config);
+    SCReturnInt(TM_ECODE_OK);
+}
+
+/**
+ * \brief This function prints stats to the screen at exit.
+ * \param tv pointer to ThreadVars
+ * \param data pointer that gets cast into DPDKThreadVars for ptv
+ */
+void ReceiveDPDKThreadExitStats(ThreadVars *tv, void *data)
+{
+    SCEnter();
+    DPDKThreadVars *ptv = (DPDKThreadVars *)data;
+
+#ifdef PACKET_STATISTICS
+    if (ptv->queue_id == 0) {
```

Comment:
does this mean the first thread prints the stats? I think its also the most likely to be killed first. Not sure if that might make it miss some packets. Probably not because there is time and sync between forcing a thread out of its capture loop and invoking this callback.

Reply:
Right, during my testing I've never had problems with stats not being printed out nor with the count of packets. Admittedly, the packet count tests had a low packet rate and ended before Suricata was shut down (so there were no transmitted packets between the stats dump and port shutdown). I believe the current solution provides good accuracy in terms of packet count.